metapool/sample_sheet.py

```diff
@@ -1488,42 +1524,99 @@ def __init__(self, path=None):
 
-    sheet = MetagenomicSampleSheetv101(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # for simplicity's sake, assume the first row marks the [Header]
+    # column and raise an Error if not. By convention it should be, once
+    # legacy comments and whitespace rows are removed.
+    if df[0][0] != '[Header]':
+        raise ValueError("Top section is not [Header]")
 
-    sheet = MetagenomicSampleSheetv100(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # identify the beginning of the following section and remove everything
+    # from the start of the first section on down. Remove the now redundant
+    # [Header] from the top row as well.
+    next_section_start = df.index[df[0].str.startswith("[")].tolist()[1]
+    df = df.iloc[1:next_section_start]
 
-    sheet = MetagenomicSampleSheetv90(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # lastly, trim off the additional all-empty columns that are now
+    # present after the removal of the other sections.
+    df.dropna(how='all', axis=1, inplace=True)
 
-    sheet = MetatranscriptomicSampleSheetv10(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # set the index to the attributes column of the sample-sheet, replacing
+    # the numeric index which now isn't needed. The dataframe will now just
+    # contain the index column and a single column named 1.
+    df.set_index(0, inplace=True)
+
+    # return the value of key '1'. This will return an immediately
+    # recognizable dictionary of key/value pairs.
+    results = df.to_dict()[1]
+
+    # conversion to dict causes SheetVersion to be wrapped in single ticks.
+    # e.g.: "'100'". These should be removed if present.
+    if 'SheetVersion' in results:
+        results['SheetVersion'] = results['SheetVersion'].replace("'", "")
+
+    return results
+
+
+def load_sample_sheet(sample_sheet_path):
+    types = [AmpliconSampleSheet, MetagenomicSampleSheetv101,
+             MetagenomicSampleSheetv100, MetagenomicSampleSheetv90,
+             AbsQuantSampleSheetv10, MetatranscriptomicSampleSheetv0,
+             MetatranscriptomicSampleSheetv10]
+
+    header = _parse_header(sample_sheet_path)
+
+    required_attributes = ['Assay', 'SheetType', 'SheetVersion']
+    missing_attributes = []
+    for attribute in required_attributes:
+        if attribute not in header:
+            missing_attributes.append(f"'{attribute}'")
+
+    if len(missing_attributes) != 0:
+        raise ValueError("The following fields must be defined in [Header]: "
+                         " %s" % ", ".join(missing_attributes))
+
+    sheet = None
+
+    for type in types:
+        m = True
+        for attribute in ['SheetType', 'SheetVersion', 'Assay']:
+            if type._HEADER[attribute] != header[attribute]:
```

Comment:
I think this is performing some type of validation. Why not pass the `header` to a member method of the sheet, something like:

```
match = False
for sheet_type in sheet_types:
    if sheet_type.matches(header):
        match = True
        break

if not match:
    raise ValueError("...")

# do stuff with sheet_type
```

Reply:
Are sample sheets enforced to have differing headers? Or does the order of evaluation of `types` matter?

Reply:
The order of evaluation of types shouldn't matter - Although we've never said it explicitly, the triplet should uniquely associate a sample-sheet with one and only one SampleSheet() class because we have SheetVersion. A sample-sheet that would otherwise pass validation as a version 100 or a version 101 should only pass one because it has to define SheetVersion as either 100 or 101.

[Headers] don't have to be different but there's nothing in our spec that says they need to remain the same. The only requirement we have is that they must all now have SheetType, SheetVersion, and Assay defined, even if they're legacy sample-sheets.

Reply:
I think this is a good idea but something I was planning on doing once we migrated to using DataFrames. It would be easy to have matches(my_header_dataframe) be in the base class and return True or False based on whether the three attributes in the frame matched the values defined in each child class.

Reply:
It seems like being defensive with perhaps a simple runtime check to assert the available set of sample sheet types are mutually exclusive would make sense.

Reply:
Respectfully, paying for that assertion every time a sample-sheet is created doesn't make that much sense to me, although I think it would be great to have as a test?

Reply:
If the consequences of violating this expectation are bad, then it is important to have assurance of expectation rather than a hope based strategy. There are a lot of ways to go about this, and I'll defer on a reasonable solution for this codebase as I don't know much about it